File: 2583.kth-largest-sum-in-a-binary-tree.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthLargestLevelSum(self, root: Optional[TreeNode], k: int) -> int:
        queue = deque([root])
        levels = []
        while queue:
            length = len(queue)
            sum = 0
            for i in range(0, length):
                node = queue.popleft()
                sum += node.val
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            levels.append(sum)
        if len(levels) < k:
            return -1
        levels.sort
        return levels[len(levels) - k]


# @leet end

File: 2583.kth-largest-sum-in-a-binary-tree.cpp
// @leet start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
#include <deque>
#include <vector>
class Solution {
public:
  long long kthLargestLevelSum(TreeNode *root, int k) {
    std::deque<TreeNode *> queue;
    queue.push_back(root);
    std::vector<long long> levels;
    while (queue.size()) {
      int length = queue.size();
      long long sum = 0;
      for (int i = 0; i < length; i++) {
        TreeNode *node = queue.front();
        sum += node->val;
        queue.pop_front();
        if (node->left)
          queue.push_back(node->left);
        if (node->right)
          queue.push_back(node->right);
      }
      levels.push_back(sum);
    }
    if (k > levels.size()) {
      return -1;
    }
    sort(levels.begin(), levels.end());
    return levels[levels.size() - k];
  }
};
// @leet end
